src/commons/api/aspa.rs

```diff
@@ -0,0 +1,370 @@
+                write!(f, ", ")?;
+            }
+            write!(f, "{}", self.providers[i])?;
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for AspaDefinition {
+    type Err = AspaConfigurationFormatError;
+
+    // example: 65000 => 65001, 65002(v4), 65003(v6)
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut parts = s.split("=>");
+
+        let customer = {
+            let customer_str = parts.next().ok_or(AspaConfigurationFormatError::CustomerMissing)?;
+            AspaCustomer::from_str(customer_str.trim())
+                .map_err(|_| AspaConfigurationFormatError::CustomerInvalid(customer_str.trim().to_string()))?
+        };
+
+        let mut providers = {
+            let mut providers = vec![];
+            let providers_str = parts.next().ok_or(AspaConfigurationFormatError::ProvidersMissing)?;
+            let provider_parts = providers_str.split(',');
+            for provider_part in provider_parts {
+                let provider = ProviderAs::from_str(provider_part.trim())
+                    .map_err(|_| AspaConfigurationFormatError::ProviderInvalid(provider_part.trim().to_string()))?;
+                providers.push(provider);
+            }
+            providers
+        };
+
+        // unexpected extra bits are not acceptable
+        if parts.next().is_some() {
+            Err(AspaConfigurationFormatError::ExtraParts)
+        } else {
+            // Ensure that the providers are sorted, there is at least one, and there are no duplicates
+            providers.sort_by_key(|p| p.provider());
+
+            let mut last_seen = providers
+                .first()
+                .ok_or(AspaConfigurationFormatError::ProvidersMissing)?;
+
+            if providers.len() > 1 {
+                for i in 1..providers.len() {
+                    let next = providers.get(i).unwrap(); // safe i max == .len() - 1
+                    if next.provider() == last_seen.provider() {
+                        return Err(AspaConfigurationFormatError::ProviderDuplicateAs(*last_seen, *next));
+                    }
+                    last_seen = next;
+                }
+            }
+
+            Ok(AspaDefinition::new(customer, providers))
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum AspaConfigurationFormatError {
+    CustomerMissing,
+    CustomerInvalid(String),
+    ProvidersMissing,
+    ProviderInvalid(String),
+    ProviderDuplicateAs(ProviderAs, ProviderAs),
+    ExtraParts,
+}
+
+impl fmt::Display for AspaConfigurationFormatError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "ASPA configuration format invalid: ")?;
+        match self {
+            AspaConfigurationFormatError::CustomerMissing => write!(f, "customer AS missing"),
+            AspaConfigurationFormatError::CustomerInvalid(s) => write!(f, "cannot parse customer AS: {}", s),
+            AspaConfigurationFormatError::ProvidersMissing => write!(f, "providers missing"),
```

Comment:
In the other three messages you don't include AS in the variant name but you do mention AS in the message, but for this message you do not mention AS. For consistency should you also mention AS here?

Reply:
fixed